### ISSUE TITLE
Adaptive GPU Assignment

### ISSUE THREAD

**Issue by mpoemsl** [pytorch/serve#2293]
### 🚀 The feature

Assign GPUs on worker thread creation according to current GPU status (free memory, utilization) and recent GPU failure history. This behavior can be configured with additional hyperparameters.

The feature is optional and backwards compatible in the sense that it can be disabled via a hyperparameter in the configuration.

### Motivation, pitch

Currently, worker threads are assigned GPUs upon creation in a round-robin fashion. This is desirable for many use cases and should definitely remain default behavior. However, some users might have heterogeneous GPU clusters or individual GPUs that are utilized by other processes. Additionally, some GPUs might be subject to temporary failure. An adaptive GPU manager could take this into account when assigning GPUs to workers.

### Additional context

I implemented this feature as part of my work at @textshuttle in a manner customized to their products' needs. If there is interest in this feature, I could create a more general PR for this feature. Note that users that do not need this feature can disable it via a hyperparameter in the configuration.

**Comment by mpoemsl** [comment]
Glad to hear there is interest in this! I will make a PR soon.

Not sure how much I am allowed to say about my work at TextShuttle, but we're using `TorchServe` in production to deploy multiple models and we identified a need for this feature in our setup.

@lxning did you also have a chance to look at #2292? Would you say that this feature is also worth making a PR?

**Comment by lxning** [comment]
@mpoemsl yes, [2292](https://github.com/pytorch/serve/issues/) is a nice optional feature. Thanks.

**Comment by mpoemsl** [comment]
@lxning Just pinging you to let you know I implemented #2292 in #2347. I would appreciate it if you could give me some feedback, so I don't make the same mistakes in the other feature PRs for #2291 and #2293.

> @mpoemsl yes, [2292](https://github.com/pytorch/serve/issues/) is a nice optional feature. Thanks.
